contracts/Safe.sol

```diff
@@ -444,8 +400,41 @@ contract Safe is
         address gasToken,
         address refundReceiver,
         uint256 _nonce
-    ) public view override returns (bytes32) {
-        return keccak256(encodeTransactionData(to, value, data, operation, safeTxGas, baseGas, gasPrice, gasToken, refundReceiver, _nonce));
+    ) public view override returns (bytes32 txHash) {
+        bytes32 separatorHash = domainSeparator();
+        /* solhint-disable no-inline-assembly */
+        assembly {
+            let ptr := mload(0x40)
+
+            calldatacopy(ptr, data.offset, data.length)
+            let calldataHash := keccak256(ptr, data.length)
+
+            // Prepare the SafeTX struct for hashing, overwriting the copied calldata
+            mstore(ptr, SAFE_TX_TYPEHASH)
+            mstore(add(ptr, 32), to)
+            mstore(add(ptr, 64), value)
+            mstore(add(ptr, 96), calldataHash)
+            mstore(add(ptr, 128), operation)
+            mstore(add(ptr, 160), safeTxGas)
+            mstore(add(ptr, 192), baseGas)
+            mstore(add(ptr, 224), gasPrice)
+            mstore(add(ptr, 256), gasToken)
+            mstore(add(ptr, 288), refundReceiver)
+            mstore(add(ptr, 320), _nonce)
+
+            // Hash the SafeTX struct and store it at the end of the result
+            // Hashing first so we can re-use the same memory block for the result
+            mstore(add(ptr, 34), keccak256(ptr, 352))
+            // Store the EIP-712 prefix (0x1901), note that strings are right-padded.
+            // We write it before the domain separator and hash to use the remaining space.
+            mstore(ptr, "\x19\x01")
+            // Store the domain separator
+            mstore(add(ptr, 2), separatorHash)
+
+            // Calculate the hash
+            txHash := keccak256(ptr, 66)
```

Comment:
Just for completeness, can we compare with something like this:

```suggestion
            // Hash the SafeTX struct and store it at the end of the result
            // Hashing first so we can re-use the same memory block for the result
            mstore(add(ptr, 64), keccak256(ptr, 352))
            // Store the EIP-712 prefix (0x1901), note that integers are left-padded
            // so the EIP-712 encoded data starts at add(ptr, 30).
            mstore(ptr, 0x1901)
            // Store the domain separator
            mstore(add(ptr, 32), separatorHash)

            // Calculate the hash
            txHash := keccak256(add(ptr, 30), 66)
```

Reply:
Rationale: Solidity won't emit a `push32 0x19010000..00` but instead a `push2 0x1901` at the cost of an extra addition.

Reply:
Gas:
```
  ERC20 - transfer
           Used 51800n gas for >transfer<
    ✔ with an EOA (73ms)
           Used 82500n gas for >transfer<
    ✔ with a single owner Safe
           Used 88381n gas for >transfer<
    ✔ with a single owner and guard Safe
           Used 89553n gas for >transfer<
    ✔ with a 2 out of 2 Safe
           Used 96607n gas for >transfer<
    ✔ with a 3 out of 3 Safe
           Used 96595n gas for >transfer<
    ✔ with a 3 out of 5 Safe
```

codesize:

```
Safe 21502 bytes (limit is 24576)
SafeL2 22282 bytes (limit is 24576)
```